classification/feature_generation/feature_utilities.py

```diff
@@ -0,0 +1,373 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import datetime
+import pytz
+import logging
+import numpy as np
+import time
+
+
+EPOCH_DT = datetime.datetime(1970, 1, 1, tzinfo=pytz.utc)
+
+
+def np_pad_repeat_slice(slice, window_size):
+    """ Pads slice to the specified window size.
+
+  Series that are shorter than window_size are repeated into unfilled space.
+
+  Args:
+    slice: a numpy array.
+    window_size: the size the array must be padded to.
+
+  Returns:
+    a numpy array of length window_size in the first dimension.
+  """
+
+    slice_length = len(slice)
+    assert (slice_length <= window_size)
+    reps = int(np.ceil(window_size / float(slice_length)))
+    return np.concatenate([slice] * reps, axis=0)[:window_size]
+
+def np_zero_pad_slice(slice, window_size, random_state):
+    """ Pads slice to the specified window size.
+
+  Series that are shorter than window_size are repeated into unfilled space.
+
+  Args:
+    slice: a numpy array.
+    window_size: the size the array must be padded to.
+
+  Returns:
+    a numpy array of length window_size in the first dimension.
+  """
+
+    slice_length = len(slice)
+    delta = window_size - slice_length
+    assert delta >= 0
+    _, n_feat = slice.shape
+    padded = np.zeros([window_size, n_feat], dtype=slice.dtype)
+    offset = random_state.randint(0, delta + 1)
+    return np.concatenate([slice] * reps, axis=0)[offset:offset+window_size]
+
+def np_pad_repeat_slice_2(slice, window_size, random_state):
+    """ Pads slice to the specified window size.
+
+  Series that are shorter than window_size are repeated into unfilled space.
+
+  Args:
+    slice: a numpy array.
+    window_size: the size the array must be padded to.
+
+  Returns:
+    a numpy array of length window_size in the first dimension.
+  """
+
+    slice_length = len(slice)
+    delta = window_size - slice_length
+    assert delta >= 0
+    slice = slice.copy()
+    GAP_LOGDT = 100
+    slice[0, 1] = GAP_LOGDT
+    reps = int(np.ceil(window_size / float(slice_length)))
+    repeated = np.concatenate([slice] * reps, axis=0)
+    offset = random_state.randint(0, window_size)
+    return np.roll(repeated, offset, axis=0)[:window_size]
+
+def np_array_random_fixed_length_extract(random_state, input_series,
+                                         output_length):
+
+    cropped = input_series[start_offset:end_offset]
+
+    return np_pad_repeat_slice_2(cropped, output_length, random_state)
+
+
+def empty_data(window_size, series):
+    num_features_inc_timestamp = series.shape[1]
+    return (np.empty(
+        [0, 1, window_size, num_features_inc_timestamp - 1],
+        dtype=np.float32), np.empty(
+            shape=[0, window_size], dtype=np.int32), np.empty(
+                shape=[0, 2], dtype=np.int32), np.empty(
+                    shape=[0], dtype=str))     
+
+
+def cook_features(features, id_):
```

Comment:
Description of what this does? :-)

Reply:
It cooks the features so that they are no longer raw ;-)

(I will add something).